File: src/website/validators.js
/**
 * Validation utilities for Goal Pyramid frontend
 * Provides reusable validation functions matching backend validation
 */

// Validation constants (must match backend)
const DISPLAY_NAME_MIN_LENGTH = 2;
const DISPLAY_NAME_MAX_LENGTH = 20;
const BIO_MAX_LENGTH = 500;

function validateDisplayName(displayName) {
  if (!displayName || !displayName.trim()) {
    return { isValid: false, error: 'Display name is required' };
  }

  const name = displayName.trim();

  if (name.length < DISPLAY_NAME_MIN_LENGTH) {
    return { isValid: false, error: `Display name must be at least ${DISPLAY_NAME_MIN_LENGTH} characters` };
  }

  if (name.length > DISPLAY_NAME_MAX_LENGTH) {
    return { isValid: false, error: `Display name must not exceed ${DISPLAY_NAME_MAX_LENGTH} characters` };
  }

  return { isValid: true, error: null };
}

function validateBio(bio) {
  if (!bio) {
    return { isValid: true, error: null };
  }

  if (bio.length > BIO_MAX_LENGTH) {
    return { isValid: false, error: `Bio must not exceed ${BIO_MAX_LENGTH} characters` };
  }

  return { isValid: true, error: null };
}

function validateProfileData(displayName, bio) {
  let result = validateDisplayName(displayName);
  if (!result.isValid) {
    return result;
  }

  result = validateBio(bio);
  if (!result.isValid) {
    return result;
  }

  return { isValid: true, error: null };
}

function getValidationConstants() {
  return {
    DISPLAY_NAME_MIN_LENGTH,
    DISPLAY_NAME_MAX_LENGTH,
    BIO_MAX_LENGTH
  };
}

File: src/api/utils/validators.py
DISPLAY_NAME_MIN_LENGTH = 2
DISPLAY_NAME_MAX_LENGTH = 20
BIO_MAX_LENGTH = 500


def validate_display_name(display_name):
    if not display_name or not display_name.strip():
        return (False, 'display_name is required')
    
    name = display_name.strip()
    
    if len(name) < DISPLAY_NAME_MIN_LENGTH:
        return (False, f'display_name must be at least {DISPLAY_NAME_MIN_LENGTH} characters')
    
    if len(name) > DISPLAY_NAME_MAX_LENGTH:
        return (False, f'display_name must not exceed {DISPLAY_NAME_MAX_LENGTH} characters')
    
    return (True, None)


def validate_bio(bio):
    if bio is None:
        return (True, None)
    
    if len(bio) > BIO_MAX_LENGTH:
        return (False, f'bio must not exceed {BIO_MAX_LENGTH} characters')
    
    return (True, None)


def validate_profile_data(display_name, bio=None, for_update=False):
    if for_update and not display_name:
        pass  # Skip display_name validation if empty during update
    else:
        is_valid, error = validate_display_name(display_name)
        if not is_valid:
            return (is_valid, error)
    
    is_valid, error = validate_bio(bio)
    if not is_valid:
        return (is_valid, error)
    
    return (True, None)

